docs/route.md

```diff
@@ -490,9 +490,13 @@ In example 2, since `Layout` is the top-level component in both routes, the DOM
 
 ---
 
-#### Authentication
+#### Redirection
 
-The RouterResolver's `onmatch` hook can be used to run logic before the top level component in a route is initializated. The example below shows how to implement a login wall that prevents users from seeing the `/secret` page unless they login.
+The RouterResolver's `onmatch` hook can be used to run logic before the top level component in a route is initializated. You can use either `m.route.set()` or `window.history.back()` `.forward()`, etc... When redirecting with the `history` API, the `onmatch` hook must return `new Promise(function(){})` to suspend the resolution of the current route. A fresh Promise must be returned every time since recycling a pending Promise leads to memory leaks. `m.route.set()` takes care of cancelling the resolution of the aborted route, there's no need to return a pending Promise in that case.
```

Comment:
@pygy 

> It isn't it is a limitation inherent to Promises.

I see what you mean now - it's adding a handler never to be executed. But my question is this: how would a shared `new Promise(() => {})` not get collected? [Here's the dependency chain I'm seeing](https://github.com/lhorie/mithril.js/blob/next/api/router.js#L32-L34):

```
original's `resolve`:
  closure: original

original's `reject`:
  closure: original

original: new Promise((resolve, reject) => {})
  closure: none
  subtasks: [
    resolve: wrapper's resolve
    reject: wrapper's reject
  ]

wrapper: Promise.resolve(original).then(...)
  closure: none
  subtasks: [
    resolve: function(resolved) { update(payload, resolved) }
    rejects: bail
  ]
```

- `wrapper` stores no reference to `original`. It just invokes `original.then(resolve, reject)`, where the two callbacks are native closures that reference `wrapper`.
- `original` stores references to `wrapper`, added by `original.then(resolve, reject)`.

Even a simple reference counting GC can properly free this chain without memory leaks when both `original`'s `resolve` and `reject` are fully dereferenced. So memory leaks aren't a concern here, as long as nothing is scheduled. `new Promise(() => {})` would literally be marked as collectable as soon as the function exits.

And if there's ever a memory leak in this sequence, it's a bug in either Mithril or the engine, requiring a fix on our end either way. Or, in other words, the "you must return a newly created one to avoid memory leaks" should probably be removed entirely.

> What about [...] [insert *pending*]

Yeah, "pending" was implied there, but I'm okay with making it explicit.

> actually, returning {then(){}} works as well, and that value can be cached. Maybe we should save that as m.route.suspend?

See my above comment for why you can still cache pending promises, and why, similarly, this also works.